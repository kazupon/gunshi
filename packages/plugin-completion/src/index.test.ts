import { exec } from 'node:child_process'
import { expect, test } from 'vitest'

function runCommand(command: string): Promise<string> {
  return new Promise((resolve, reject) => {
    exec(command, (error, stdout, stderr) => {
      if (error) {
        reject(stderr)
      } else {
        resolve(stdout)
      }
    })
  })
}

const SCRIPT = `pnpx tsx packages/plugin-completion/examples/demo.node.ts complete`

test('termination', async () => {
  const output = await runCommand(`${SCRIPT} --`)
  expect(output).toMatchSnapshot()
})

test.todo('default command option', async () => {
  const output = await runCommand(`${SCRIPT} --config --`)
  expect(output).toMatchSnapshot()
})

test('subcommand only', async () => {
  const output = await runCommand(`${SCRIPT} -- dev`)
  expect(output).toMatchSnapshot()
})

test('subcommand with long option', async () => {
  const output = await runCommand(`${SCRIPT} -- dev --port`)
  expect(output).toMatchSnapshot()
})

test('subcommand with short option', async () => {
  const output = await runCommand(`${SCRIPT} -- dev -H`)
  expect(output).toMatchSnapshot()
})

test.todo('subcommand with long option and value', async () => {
  const output = await runCommand(`${SCRIPT} -- dev --port=3`)
  expect(output).toMatchSnapshot()
})
