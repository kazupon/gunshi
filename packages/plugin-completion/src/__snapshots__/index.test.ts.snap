// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`subcommand only 1`] = `
"dev	Start dev server
:4
"
`;

exports[`subcommand with long option 1`] = `
"--port	Specify port
:4
"
`;

exports[`subcommand with long option and value 1`] = `
"completion command context [Object: null prototype] {
  name: 'complete',
  description: 'Generate shell completion script',
  omitted: false,
  callMode: 'subCommand',
  env: [Object: null prototype] {
    name: 'vite',
    description: 'Vite CLI',
    version: '0.0.0',
    cwd: undefined,
    usageSilent: false,
    subCommands: Map(5) {
      'dev' => [Object],
      'build' => [Object],
      'lint' => [Object],
      '' => [Object],
      'complete' => [Object]
    },
    leftMargin: 2,
    middleMargin: 10,
    usageOptionType: false,
    usageOptionValue: true,
    renderHeader: [Function: renderer],
    renderUsage: [Function: renderer],
    renderValidationErrors: [Function: renderer],
    plugins: [ [AsyncFunction] ]
  },
  args: [Object: null prototype] {
    help: [Object: null prototype] {
      type: 'boolean',
      short: 'h',
      description: 'Display this help message'
    },
    version: [Object: null prototype] {
      type: 'boolean',
      short: 'v',
      description: 'Display this version'
    }
  },
  values: [Object: null prototype] {},
  positionals: [ 'complete' ],
  rest: [ 'dev', '--port=3' ],
  _: [ 'complete', '--', 'dev', '--port=3' ],
  tokens: [
    { kind: 'positional', index: 0, value: 'complete' },
    { kind: 'option-terminator', index: 1 },
    { kind: 'positional', index: 2, value: 'dev' },
    { kind: 'positional', index: 3, value: '--port=3' }
  ],
  toKebab: undefined,
  log: [Function: log],
  validationError: undefined,
  extensions: [Object: null prototype] {
    'g:global': {
      showVersion: [Function: showVersion],
      showHeader: [AsyncFunction: showHeader],
      showUsage: [AsyncFunction: showUsage],
      showValidationErrors: [AsyncFunction: showValidationErrors]
    },
    'g:renderer': {
      text: [AsyncFunction: text],
      loadCommands: [AsyncFunction: loadCommands]
    },
    'g:completion': undefined
  }
}
:4
"
`;

exports[`subcommand with short option 1`] = `
"-H	Specify hostname
:4
"
`;

exports[`termination 1`] = `
"dev	Start dev server
build	Build project
lint	Lint project
:4
"
`;
